File: apps/api/src/services/membership_tiers/tiers.py
"""
MembershipTier CRUD service functions.

All endpoints are superadmin-only. Superadmin enforcement is handled
at the router layer via the require_superadmin dependency.
"""

import logging
from datetime import datetime

from fastapi import HTTPException, status
from sqlalchemy import func
from sqlmodel import Session, select

from src.db.membership_tiers import (
    MembershipTier,
    MembershipTierCreate,
    MembershipTierRead,
    MembershipTierUpdate,
)
from src.db.user_memberships import UserMembership

logger = logging.getLogger(__name__)


def create_tier(tier_data: MembershipTierCreate, db_session: Session) -> MembershipTierRead:
    """Create a new membership tier. Raises 409 if slug is already taken."""
    # Check slug uniqueness
    existing = db_session.exec(
        select(MembershipTier).where(MembershipTier.slug == tier_data.slug)
    ).first()
    if existing:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"A tier with slug '{tier_data.slug}' already exists",
        )

    tier = MembershipTier(
        name=tier_data.name,
        slug=tier_data.slug,
        description=tier_data.description,
        is_active=tier_data.is_active,
        priority=tier_data.priority,
        permissions=tier_data.permissions,
        creation_date=str(datetime.now()),
        update_date=str(datetime.now()),
    )

    db_session.add(tier)
    db_session.commit()
    db_session.refresh(tier)

    return MembershipTierRead.model_validate(tier)


def get_all_tiers(db_session: Session) -> list[MembershipTierRead]:
    """Return all membership tiers ordered by priority descending."""
    tiers = db_session.exec(
        select(MembershipTier).order_by(MembershipTier.priority.desc())  # type: ignore[union-attr]
    ).all()
    return [MembershipTierRead.model_validate(t) for t in tiers]


def get_tier_by_id(tier_id: int, db_session: Session) -> MembershipTierRead:
    """Return a single tier or raise 404."""
    tier = db_session.exec(
        select(MembershipTier).where(MembershipTier.id == tier_id)
    ).first()

    if not tier:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Membership tier not found",
        )

    return MembershipTierRead.model_validate(tier)


def update_tier(
    tier_id: int, tier_data: MembershipTierUpdate, db_session: Session
) -> MembershipTierRead:
    """Partially update a tier. Raises 409 if new slug conflicts."""
    tier = db_session.exec(
        select(MembershipTier).where(MembershipTier.id == tier_id)
    ).first()

    if not tier:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Membership tier not found",
        )

    # If slug is being changed, check uniqueness
    if tier_data.slug is not None and tier_data.slug != tier.slug:
        existing = db_session.exec(
            select(MembershipTier).where(MembershipTier.slug == tier_data.slug)
        ).first()
        if existing:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=f"A tier with slug '{tier_data.slug}' already exists",
            )

    update_data = tier_data.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(tier, key, value)

    tier.update_date = str(datetime.now())

    db_session.add(tier)
    db_session.commit()
    db_session.refresh(tier)

    return MembershipTierRead.model_validate(tier)


def get_tier_user_count(tier_id: int, db_session: Session) -> dict:
    """Return the count of active user memberships for a given tier."""
    # Verify tier exists first
    tier = db_session.exec(
        select(MembershipTier).where(MembershipTier.id == tier_id)
    ).first()

    if not tier:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Membership tier not found",
        )

    count = db_session.exec(
        select(func.count(UserMembership.id))  # type: ignore[arg-type]
        .where(UserMembership.tier_id == tier_id)
        .where(UserMembership.status == "active")
    ).one()

    return {"tier_id": tier_id, "active_user_count": count}

File: apps/web/services/membership_tiers/tiers.ts
import { getAPIUrl } from '@services/config/config'
import { RequestBodyWithAuthHeader, getResponseMetadata } from '@services/utils/ts/requests'

/*
  Membership Tiers service
  - GET    tiers/          — list all tiers
  - POST   tiers/          — create tier
  - GET    tiers/{id}      — get tier by id
  - PUT    tiers/{id}      — update tier
  - GET    tiers/{id}/count — user count for tier
*/

export type TierPermissions = Record<string, boolean | number | string>

export type CreateTierBody = {
  name: string
  slug: string
  description?: string
  priority?: number
  is_active?: boolean
  permissions?: TierPermissions
  stripe_product_id?: string
  stripe_price_monthly_id?: string
  stripe_price_yearly_id?: string
  price_monthly_display?: string
  price_yearly_display?: string
}

export type UpdateTierBody = Partial<CreateTierBody>

export type MembershipTier = {
  id: number
  name: string
  slug: string
  description: string | null
  priority: number
  is_active: boolean
  permissions: TierPermissions
  creation_date: string
  update_date: string
  stripe_product_id?: string | null
  stripe_price_monthly_id?: string | null
  stripe_price_yearly_id?: string | null
  price_monthly_display?: string | null
  price_yearly_display?: string | null
}

export async function getTiers(access_token?: string): Promise<MembershipTier[]> {
  const result = await fetch(
    `${getAPIUrl()}tiers/`,
    RequestBodyWithAuthHeader('GET', null, null, access_token)
  )
  const res = await getResponseMetadata(result)
  return res.data
}

export async function getTier(tierId: number | string, access_token?: string): Promise<MembershipTier> {
  const result = await fetch(
    `${getAPIUrl()}tiers/${tierId}`,
    RequestBodyWithAuthHeader('GET', null, null, access_token)
  )
  const res = await getResponseMetadata(result)
  return res.data
}

export async function createTier(body: CreateTierBody, access_token: string) {
  const result = await fetch(
    `${getAPIUrl()}tiers/`,
    RequestBodyWithAuthHeader('POST', body, null, access_token)
  )
  const res = await getResponseMetadata(result)
  return res
}

export async function updateTier(tierId: number | string, body: UpdateTierBody, access_token: string) {
  const result = await fetch(
    `${getAPIUrl()}tiers/${tierId}`,
    RequestBodyWithAuthHeader('PUT', body, null, access_token)
  )
  const res = await getResponseMetadata(result)
  return res
}

export async function getTierUserCount(tierId: number | string, access_token?: string): Promise<number> {
  const result = await fetch(
    `${getAPIUrl()}tiers/${tierId}/count`,
    RequestBodyWithAuthHeader('GET', null, null, access_token)
  )
  const res = await getResponseMetadata(result)
  // API returns { count: number } or just a number
  return typeof res.data === 'number' ? res.data : (res.data?.count ?? 0)
}
